### PR TITLE
Concept docs for caching mechanisms

### PR DESCRIPTION
**Pull request by ikas**
Related to [GTC-731](https://gfw.atlassian.net/browse/GTC-731).

### CODE REVIEW

File: source/includes/api/_before_you_get_started.md
Comment (tiagojsag) [inline source/includes/api/_before_you_get_started.md]: Caching should not be though of as a way to achieve improved performance. Caching should be primarily a scaling tool. Relying on caching for performance is acknowledging we are ok with sacrificing performance for a substantial part of our requests, especially the custom ones.

File: source/includes/api/_before_you_get_started.md
Comment (tiagojsag) [inline source/includes/api/_before_you_get_started.md]: the fact that we use fastly is an implementation detail, and we shouldn't really mention it here. Perhaps it would make sense on a dev docs section, but here it's not relevant IMO. Instead, what we can do is briefly explain how an HTTP cache works: when a request comes in, it computes the response, saves it, and subsequent requests are then replied with the save response, without the need for recomputing it. I doubt anyone reading this does not know how a cache works, so it may be redundant, but I feel like that's more important background info for an API user to have.

File: source/includes/api/_before_you_get_started.md
Comment (tiagojsag) [inline source/includes/api/_before_you_get_started.md]: what do you mean by this? I'm assuming you mean requests with authentication cookies, right?

File: source/includes/api/_before_you_get_started.md
Comment (tiagojsag) [inline source/includes/api/_before_you_get_started.md]: implementation details, not relevant

File: source/includes/api/_before_you_get_started.md
Comment (tiagojsag) [inline source/includes/api/_before_you_get_started.md]: Also implementation details, not relevant

**Comment by tiagojsag** [comment]
I know this is WIP, so you probably had this written down as stuff you want to improve, but my 2 cents:
- the user docs should basically describe when the api caches and when it clears cache.
- we should NOT mention fastly (implementation detail, not important for consumers)
- we should NOT mention our magic headers, unless they somehow are visible to outside users
- we should mention and headers that may help a user debug anything - I dont know if we have any at all, but for example, if we have a header that shows a response was served by the cache, that would be useful to the end user
- our cache is nothing more than a standard public HTTP cache, so we may wanna mention that, along with a link to a useful page (wikipedia maybe?) that covers the standard headers for HTTP caching. Again, just link to it, we don't need to document it.

File: source/includes/api/_before_you_get_started.md
Comment (ikas) [reply source/includes/api/_before_you_get_started.md]: I meant responses from auth/user-related endpoints - I felt like it was a bit on the extra-safe side to highlight that auth/user information is never cached. I've rephrased so it's clearer, lmk what you think

File: source/includes/api/_before_you_get_started.md
Comment (ikas) [reply source/includes/api/_before_you_get_started.md]: Agreed - we have cache implementation details on the developer docs, so I removed redundant information from here and added a reference to the developer docs :+1:

File: source/includes/concept/_overview.md
Comment (tiagojsag) [inline source/includes/concept/_overview.md]: cache is opt-in for devs. The docs here are targeted at users, so it's not opt in. What I think you are trying to communicate is that not all endpoints have caching enabled for them.

**Comment by ikas** [comment]
> we should mention and headers that may help a user debug anything - I dont know if we have any at all, but for example, if we have a header that shows a response was served by the cache, that would be useful to the end user

we have some cache-related headers in the response of "cache-compliant" API resources, but I don't think they are being correctly populated - for instance, every hit to the dataset GET endpoint returns a `x-cache: MISS` header, never a cache hit. I'm not 100% it would be safe to refer to these headers as debugging tools, given the circumstances, but I do agree that it would be important to refer to this in the docs (if it was working as expected). Any ideas on how to move forward?

File: source/includes/concept/_overview.md
Comment (tiagojsag) [inline source/includes/concept/_overview.md]: you may also want to point out that results of authenticated GET calls are not stored - and please confirm that's the case, and if it's not, we have to fix it.

Comment (tiagojsag) [inline source/includes/concept/_overview.md]: The default cache time to live (TTL) applied to the responses stored in the RW API's cache is 3 days, but specific endpoints may specify a different expiration time for their responses.

**Review by tiagojsag — CHANGES_REQUESTED** [review]
minor stuff, otherwise LGTM

File: source/includes/concept/_overview.md
Comment (ikas) [reply source/includes/concept/_overview.md]: added and confirmed this is indeed the case

**Comment by tiagojsag** [comment]
> 
> 
> > we should mention and headers that may help a user debug anything - I dont know if we have any at all, but for example, if we have a header that shows a response was served by the cache, that would be useful to the end user
> 
> we have some cache-related headers in the response of "cache-compliant" API resources, but I don't think they are being correctly populated - for instance, every hit to the dataset GET endpoint returns a `x-cache: MISS` header, never a cache hit. I'm not 100% it would be safe to refer to these headers as debugging tools, given the circumstances, but I do agree that it would be important to refer to this in the docs (if it was working as expected). Any ideas on how to move forward?

if they are not reliable, I wouldn't mention them.

**Review by ethanroday — CHANGES_REQUESTED** [review]
Maybe the behavior is too microservice-specific, but I think this could use some fleshing out. As an API user, I'm looking for:
 - What kinds of responses will be cached? (this seems to be covered)
 - How do I know if I'm getting a cached response?
 - How do I bypass the cache on a given request?
 - How do I force the cache to refresh for a given resource?

Also one clarification question:
 - When you say authenticated GET requests are never cached, does that include requests for public resources that just happen to pass a JWT?

**Comment by tiagojsag** [comment]
> 
>     * How do I know if I'm getting a cached response?
There seems to be a response header added by Fastly that covers this. I asked Henrique not to document this because a) I'm not really sure if this is meant to be a Fastly production setting, or some dev option we have enabled for whatever reason and b) this only accounts for Fastly, not other HTTP caches (public or private) that may exist between us and the user.
> 
>     * How do I bypass the cache on a given request?
> 
>     * How do I force the cache to refresh for a given resource?
> 
I would not document these. Anyone familiar with HTTP caches knows how to bypass them, but documenting it makes it "officially supported" behavior, which is the opposite of what we want - we want users to use the cached responses whenever they are available. As for forcefully triggering a cache refresh, that's not possible, nor should it be - a malicious user could spam said endpoint to keep our caches empty, just for fun.

 
> Also one clarification question:
> 
>     * When you say authenticated GET requests are never cached, does that include requests for public resources that just happen to pass a JWT?

Yes. HTTP cache standard behavior is to not cache requests that contain authentication data (cookies, for example) because they may contain more data than anonymous requests, and thus lead to unintended info disclosure. For example, you can list datasets and include authors emails, but only if you are an ADMIN. So, if we were to cache that, anyone could retrieve that same response even without meeting those permission restrictions.

**Comment by ikas** [comment]
Hey Ethan, thanks for the feedback :) a few thoughts adding to what Tiago commented:

> * How do I know if I'm getting a cached response?

https://docs.fastly.com/en/guides/understanding-cache-hit-and-miss-headers-with-shielded-services

The link above is Fastly docs on how to understand if a request was a HIT or MISS (and it seems to be a simple and helpful explanation), so I think we can point users there, regarding detecting if users are being given a cached response. Lmk if you think this is enough, or if you prefer that we write our own explanation.

> * How do I bypass the cache on a given request?

About this, we do have a "trick in our sleeves", which is using `https://aws-production.resourcewatch.org` as the host - using this host, users completely bypass Fastly and go directly to the k8s cluster. However, I must highlight that I 100% agree with Tiago that we should not document this in user-faced docs. IMO, teaching users how to bypass cache completely defeats the overall purpose of caching. 

What we could do, however, is add some of these details that we've been discussing to the caching developer docs. To me, it seems a more appropriate place for things "how to bypass cache" :)

**Comment by ikas** [comment]
@ethanroday just updated the PR:

* Added a small section on detecting a cached response, complemented with the Fastly docs link I shared yesterday.
* Added another section on refreshing/invalidating cached responses (we don't officially support it, but there are helpful workarounds described there).

I did not add any information about bypassing cache yet, since both me and Tiago have raised some concerns about disclosing that information to public users, but I'll leave the final call to you - if you think it's something that belongs here in the API user docs, let me know and I'll add a paragraph on that :)

Re-requesting review :rocket:

**Comment by ethanroday** [comment]
Thanks for the detailed and insightful comments, @ikas and @tiagojsag. The Fastly info is helpful. Based on Tiago's suggestion, let's move the cache invalidation info that @ikas just added to the developer docs if it's not already there and leave it out of the public-facing docs.

**Comment by ethanroday** [comment]
I'll approve this now so we don't have to do another back and forth, but @ikas please do make that change before you merge.

**Review by ethanroday — APPROVED** [review]

**Comment by tiagojsag** [comment]
On cache invalidation:
- IMO cache invalidation should always happen automatically. I don't see a use case for users to invalidate cache manually.
- Users (and devs, by extension) need to know that this automatic invalidation exists, and that users shouldn't need to worry about how it happens - it just happens. That's what I tried to clarify in my latest changes.
- Devs already had a few bullet points on how they can use specific headers to cache and uncache content - this, combined with the info above, defines the expected behavior, and which tools are available for devs to implement it.
- Cache invalidation as a standalone action is not implemented on the API anywhere. Henrique's suggestion was a hack that piggybacks off that uncache header being used in certain endpoints, but I'm not comfortable at all with giving that as a "clear cache" solution. A dev that wants to clear a cache should have read about the uncache header. They should be able to connect the dots from there, if they REALLY need to clear a cache manually (which I suspect will probably won't happen - I myself never had to do it).

On dev docs:
- Added a link to the CT Fastly plugin to the dev docs. Since it will be replaced soon (TBD with what exactly) I didn't want to spend too much time or effort on that, and I feel like whatever implementation we roll out after CT will be similar-ish to this anyways, so we can tweak it then.

I also added a few lines on 3rd party caches and HTTPS, to clarify what I suspect may be frequent questions by users

@ethanroday I know you asked us to go ahead and merge without your review, but since I made some substantial changes here, I'd like your feedback on them, to make sure you agree with my course of action on this.

**Comment by ethanroday** [comment]
@tiagojsag I still approve 👍 
One small thing - I don't see a difference between ### and #### headers when I run the site locally, which makes this section kind of confusing. Unless there's a specific reason to keep them as headers, can we just turn the #### headers into bold lines?

**Pull request merged**
